### ISSUE TITLE
Implement normalized betweenness

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1797]
**What is the feature or improvement you would like to see?**
The betweenness of a vertex `v` is the number of `(s,t)` vertex pairs (`s != t != v`) whose connecting shortest path passes through `v`. When there is more than one shortest path between `(s,t)`, we take the fraction which passes through `v`.

Thus, the maximum possible betweenness value is the number of `(s,t)` pairs excluding `v`, i.e. `(V-1)(V-2)/2` where `V` is the vertex count. There is a correction factor of 2 for directed graphs. For edge betweenness, we have `V(V-1)/2`.

The _normalized betweenness_ is the standard betweenness value divided by the maximum possible value, as computed with the formulas above.

This is a feature request to add a flag to betweenness functions to compute the normalized betweenness.

This is trivial for standard betweenness, but a bit more complicated for subset betweenness, where the number of `(s,t)` pairs depends not only on the size of the source and target sets, but also on the size of their overlaps, as well as on the specific vertex we consider (which may or may not be part of the source/target sets). It is because of this complication that it would be useful to implement this in igraph instead of leaving it to users.

**Use cases for the feature**

The use case is to be able to compare standard betweenness with subset betweenness within the same graph.

Note: Naïvely one might assume that this normalized betwenness allows for comparisons between vertices of graphs of different sizes. This is not a straightforward thing to do because normalized betweenness not only decreases with graph size, but the rate of decrease depends on the specific graph structure (not only on vertex/edge counts.

**References**

**Comment by szhorvat** [comment]
Needs to make it into `0.10` as it is an almost-necessary continuation of the subset betweenness feature.

**Comment by ecpaim** [comment]
Hi,

First of all I want to thank you for maintaining this project, as I used it in a research I was doing last year and it was really useful.

Secondly, I got around this issue and it'd seem to me that the main question could be solved by, assuming S and T are the sizes of source and target set and v is the node we want to compute the betweeness of, `((S -1)*T)/2` if v is not in the target set, and `((S -1)*(T-1))/2` otherwise. (If we assume every minimal path between s and t must pass through v, then this division computes every couple of vertices, except the ones containing v);

I also assumed here that we can't compute the betweenness of a node that is not in the source set, if that is possible, then `(S*T)/2` and `(S*(T-1))/2` is also possible. Although looking at the implementation of `igraph_betweenness_subset` it seems that it is not possible to give values outside of the source set.

From here, there's some options. One could select the maximum value between these divisions and divide all nodes by the same constant, or choose different divisions based on the node being part of T and S. I write to see which one you think it's best for the project before I commit to write it.


Aside from that, the implementation of `igraph_betweenness_subset` has an odd behaviour: if we consider the very simple graph:

`0 - 1 - 2`

Where `0,1` is the source set and `2` is the target set, the function `igraph_betweenness_subset` returns a value of 0.5 for the betweeness of  node 1! From the definition, if betweeness of v is the number of geodesics passing through v divided by the total number of geodesics, then we would expect the betweeness of node 1 to be equal to 1 as well. And, indeed if we just change the edges to get the direct graph `0 -> 1 -> 2`, we get the expected 1. 

Why is that? Looking at the code, it is because the result for undirected graphs is always divided by 2. But, if we are comparing source to target sets there is already an implicit direction happening, and the path `2 - 1 - 0` never gets computed by the path algorithms, therefore making it unnecessary to divide by 2!

I can't say exactly if this is an error, as for all I know it could be working as intended. Although the other famous python library `networkx` gives the same 'odd  behaviour', even if it is contraditory with the [equation they posted](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.centrality.betweenness_centrality_subset.html) in the docs about this function.

**Comment by GroteGnoom** [comment]
Just to check:
This is about all the public functions in src/centrality/betweenness.c, right?
You don't want to take the cutoff into account for the maximum in the functions which use a cutoff, right?

**Comment by szhorvat** [comment]
Yes, this would take into account subset betweenness, but not any cutoff. 

To be honest, I did not think carefully about the effects of cutoff. That would be a separate topic.

**Comment by GroteGnoom** [comment]
> For edge betweenness, we have V(V-1)/2.

Shouldn't that be ⌊V<sup>2</sup> / 4⌋, as discussed here: https://math.stackexchange.com/questions/4056495/maximum-possible-edge-betweenness-of-a-graph ?

**Comment by szhorvat** [comment]
I would need more time to think this through, but here's my take on it at this moment:

When calculating betweenness, we consider shortest paths between all pairs of vertices. There are V(V-1)/2 pairs. Thus the interpretation of normalized betweenness is very intuitive: it is the _fraction_ of (pairwise) shortest paths passing through an edge or a vertex.

As you point out, the largest possible value of betweenness is not the same as the number of pairs. It is smaller than that. Yet the number we get when dividing with the largest possible value is not as intuitive as a "a fraction of `n` total".  Yes, I admit this that is subjective to some extent.

However, the reason I want normalized betweenness is so that:
 - We can use subset betweenness for estimating the complete betweenness.
 - We can compare subset betweenness values for different subsets.

Calculating the largest possible value for subset betweenness is not trivial. Calculating the number of pairs is fairly simple.  Furthermore, "what fraction of the total number of pairs" is an easier to interpret concept than "divided by the largest possible value".

For this reason, I think it's better to use the number of pairs.

----

Unfortunately, I do not have enough time these days to look into this in detail (I'm sorry, I didn't even have time to read in full the M.SE post and Misha's answer there).  In order to understand the topic better, we need to look at papers which work out the details of how to use subset betweenness (with a random subset) for estimating the complete betweenness. I think there was a paper by Brandes on this?

**Comment by szhorvat** [comment]
@ntamas, opinions?

**Comment by ntamas** [comment]
Sorry, this has slipped through the cracks somehow. Yes, I'd simply use the number of pairs as an approximation.

**Comment by GroteGnoom** [comment]
> In order to understand the topic better, we need to look at papers which work out the details of how to use subset betweenness (with a random subset) for estimating the complete betweenness. I think there was a paper by Brandes on this?

I guess we're now using the algorithm described in 'A faster algorithm for betweenness centrality' by Brandes in 2001. 

Then 'Centrality estimation in large networks' by Brandes and Pich in 2007 describes how to speed things up by using a vertex subset. This elements of this subset are called 'pivots', and a large part of this paper is about choosing them. They conclude that just a random selection of vertices work best and expect sophisticated pivot choosing algorithms to run into trouble.

But there's still issues, which are addressed in 'Better approximation of betweenness centrality' by Geisberger, Sanders and Schultes in 2008. One of the issues is that being near a pivot can lead to a huge overestimation of its betweenness:

> For example, consider a degree-two node v connecting a degree-one node u to the rest of the network. If u is selected as a pivot, the betweenness of v is overestimated by a factor of n/k.

They fix this by putting more emphasis on edges further away from a pivot. Two methods they try for this are a linear scaling, and only counting the second half of the path.

So for this issue I should probably focus on Brandes and Pich and look at the normalization there.

**Comment by GroteGnoom** [comment]
With my current schedule, I don't think I should continue on this. The gaps between working on this are too big, so it takes a lot of effort every time to start focusing on it. It will take a long time to get comfortable with this subject, while producing nothing visible in the mean time, which makes it seem like I'm not doing any work.

**Comment by szhorvat** [comment]
It would be useful to document the research you've done so far. Perhaps we should schedule a meeting sometime within the next 3 weeks and do that together.

**Comment by szhorvat** [comment]
New argument should be added in 1.0 to prepare the ground for this feature. Retargeting for 1.0 so we don't forget (but the implementation should come in 1.1 only).

**Comment by szhorvat** [comment]
Triage: Consider adding a separate function.

**Comment by DOSull** [comment]
Can you provide any more information as to how `betweenness(..., normalized = TRUE, cutoff = d)` works when cutoff is set such that shortest paths are restricted to those shorter than `d`? 

When `cutoff = -1` and all shortest paths are considered, I can see that the normalization factor is (n-1)(n-2)/2 or some close relative of it (ie the total count of all possible shortest paths), it's less clear when shortest paths are restricted in their length.

Thanks for any insights!

**Comment by szhorvat** [comment]
Normalization is not implemented in the igraph C library. I suppose you are asking about the R library. Normally I would ask to post such questions on the forum, but in this case please open an issue in https://github.com/igraph/rigraph/ so it can be discussed whether any changes are needed there.

(A quick peek at the R code suggests that it just divides by the same factor regardless whether there is a cutoff—my first reaction is that I'm not sure this is a good thing.)

**Comment by DOSull** [comment]
Thank you for this. I have a feeling that you are right that normalization should not be considered valid when a cutoff is used.
